tests/integration/test_self_monitoring.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+from pathlib import Path
+from typing import List
+import pytest
+import jubilant
+from jubilant import Juju
+from tenacity import retry, stop_after_attempt, wait_fixed
+import yaml
+from conftest import APP_NAME, COS_CHANNEL
+from pytest_bdd import given, when, then
+
+OTEL_COLLECTOR_APP_NAME = "opentelemetry-collector"
+METADATA = yaml.safe_load(Path("./charmcraft.yaml").read_text())
+# get the charm os base without the trailing `:architecture` part
+APP_BASE = next(iter(METADATA["platforms"])).split(":")[0]
+
+
+@pytest.fixture(scope="module")
+def patch_update_status_interval(juju: Juju):
+    # patch the update-status-hook-interval because if the otel collector charm handles an event, it will regenerate its config and overwrite the config patch
+    juju.model_config({"update-status-hook-interval": "1h"})
+    yield
+    juju.model_config(reset="update-status-hook-interval")
+
+
+def _patch_otel_collector_log_level(juju: Juju, unit_no=0):
+    # patch the collector's log level to INFO
+    juju.ssh(
+        f"{OTEL_COLLECTOR_APP_NAME}/{unit_no}",
+        f"sudo sed -i 's/level: WARN/level: INFO/' /etc/otelcol/config.d/{OTEL_COLLECTOR_APP_NAME}_{unit_no}.yaml",
+    )
+    # resatart the snap for the updates to take place
+    juju.ssh(f"{OTEL_COLLECTOR_APP_NAME}/{unit_no}", "sudo snap restart opentelemetry-collector")
+
+
+def _is_pattern_in_snap_logs(juju: Juju, grep_filters: List[str]):
+    cmd = (
+        "sudo snap logs opentelemetry-collector -n=all"
+        + " | "
+        + " | ".join([f"grep {p}" for p in grep_filters])
+    )
+    otelcol_logs = juju.ssh(f"{OTEL_COLLECTOR_APP_NAME}/0", command=cmd)
+
+    if not otelcol_logs:
+        raise Exception(f"Filters {grep_filters} not found in the {OTEL_COLLECTOR_APP_NAME} logs")
+    return True
```

Comment:
It's a bit odd to have a function that either raises or returns True, and you assert that it returns True.

```suggestion
def assert_pattern_in_snap_logs(juju: Juju, grep_filters: List[str]):
    cmd = (
        "sudo snap logs opentelemetry-collector -n=all"
        + " | "
        + " | ".join([f"grep {p}" for p in grep_filters])
    )
    otelcol_logs = juju.ssh(f"{OTEL_COLLECTOR_APP_NAME}/0", command=cmd)

    assert otelcol_logs, f"Filters {grep_filters} not found in the {OTEL_COLLECTOR_APP_NAME} logs"
```

this is a common pattern, to write custom assert functions.